rainier-core/src/main/scala/rainier/sampler/Variational.scala

```diff
@@ -0,0 +1,127 @@
+package rainier.sampler
+
+import rainier.compute._
+import rainier.core._
+
+trait Guide {
+  def density: Real // density of variational distributions
+  def params: Seq[Variable] // params to optimize
+  def forward: Map[Variable, Real] // map between original model params and f(params, eps)
+  def generate(implicit rng: RNG): Seq[Double]
+  def sample(paramSettings: Map[Variable, Double]): Map[Variable, Double]
+  def init: Map[Variable, Double]
+
+  private def transformDensity(targetDensity: Real): Real = {
+    Real.substituteVariable(targetDensity, forward)
+  }
+  def loss(targetDensity: Real): Real = {
+    transformDensity(targetDensity) - density
+  }
+}
+
+case class MeanFieldNormalGuide(targetVariables: Seq[Variable])(
+    implicit rng: RNG)
+    extends Guide {
+  private val normals: Map[(Variable, Variable), (Real, Continuous)] =
+    targetVariables
+      .map(tv => {
+        val mu = new Variable
+        val sigma = new Variable
+        val dist = Normal(mu, sigma)
```

Comment:
the way you deconstruct `Normal` here to reach in and grab its variable and later to use `realLogDensity` which is meant to be private feels to me like you shouldn't really be using `Normal` at all and should just be redefining the normal pdf locally to this class - it's not that complex a function.